### PR TITLE
Point only to the identifiers in the typo suggestions of shadowed names instead of the entire struct

### LINKED ISSUE #103358: Point at shadowed name binding in `you might have meant to refer to` suggestion
Makes sense, I just would rather we pointed at the ident for the item and not the whole thing.

_Originally posted by @estebank in https://github.com/rust-lang/rust/pull/103111#discussion_r1001294984_

The suggestion was introduced by #103111. The suggestion point to the definition of the suggested item. Pointing to the shadowed identifier would be more precise:
- the ident of the item in case of direct reference;
- the import in case of indirect reference.

### PR DESCRIPTION
**Pull request by zbyrn**
Fixes #103358.

As discussed in the issue, the `Span` of the candidate `Ident` for a typo replacement is stored alongside its `Symbol` in `TypoSuggestion`. Then, the span of the identifier is what the "you might have meant to refer to" note is pointed at, rather than the entire struct definition.

Comments in #103111 and the issue both suggest that it is desirable to:
1. include names defined in the same crate as the typo,
2. ignore names defined elsewhere such as in `std`, _and_
3. include names introduced indirectly via `use`.

Since a name from another crate but introduced via `use` has non-local `def_id`, to achieve this, a suggestion is displayed if either the `def_id` of the suggested name is local, or the `span` of the suggested name is in the same file as the typo itself.

Some UI tests have also been modified to reflect this change.

r? @cjgillot

### CODE REVIEW

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 1]
```diff
@@ -691,9 +691,20 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+            file1 == file2
+        };
+        // print 'you might have meant' if the candidate is (1) is a shadowed name with
+        // accessible definition and (2) either defined in the same crate as the typo
+        // (could be in a different file) or introduced in the same file as the typo
+        // (could belong to a different crate)
+        if let TypoCandidate::Shadowed(res, Some(sugg_span)) = typo_sugg
+            && res
+                .opt_def_id()
+                .map_or(false, |id| id.is_local() || is_in_same_file(span, sugg_span))
```
Comment (cjgillot) [inline compiler/rustc_resolve/src/late/diagnostics.rs:707]: Do you have an example where removing the `is_local` and `is_in_same_file` filters give the wrong suggestions?
TBH, I'm not fond of the filename-based logic. We should either formulate the logic in terms of modules, or drop `is_in_same_file`.

Comment (zbyrn) [reply compiler/rustc_resolve/src/late/diagnostics.rs:707]: When we shadow something in another crate, the suggestion would potentially point at a distant file in which the shadowed name is defined. For example, compiling
```
mod m {
    pub enum MyEnum {
        Result
    }
}

use m::MyEnum::Result;

fn foo() -> Result<u8, str> {
    unimplemented!()
}

fn main() { }
```
will give the following error message:
```
error[E0573]: expected type, found variant `Result`
  --> test.rs:9:13
   |
9  | fn foo() -> Result<u8, str> {
   |             ^^^^^^^^^^^^^^^ not a type
   |
  ::: /[REDACTED]/rust/library/std/src/prelude/v1.rs:35:24
   |
35 | pub use crate::result::Result::{self, Err, Ok};
   |                        ------ you might have meant to refer to this enum
   |
help: consider importing one of these items instead
   |
7  | use std::fmt::Result;
   |
7  | use std::io::Result;
   |
7  | use std::result::Result;
   |
7  | use std::thread::Result;
   |

error: aborting due to previous error

For more information about this error, try `rustc --explain E0573`.
```
Note that the suggestion here points at something in the implementation of `std`. If this is okay, we could just remove both checks and simplify the code. The suggestion given here is not wrong technically, but I think it is not very helpful either.

I'm not a fan of the filename hack as well, and it was certainly a last resort as I'm still familiarizing myself with the code base. Could we achieve the same effect by comparing the `ctx`'s of the two spans?

Comment (zbyrn) [reply compiler/rustc_resolve/src/late/diagnostics.rs:707]: On a second thought, we could actually use the span of the resolver's module span (`self.r.graph_root.span`) to test whether the underlined definition is in the current module. Pushing a change and amending this PR shortly

Comment (cjgillot) [reply compiler/rustc_resolve/src/late/diagnostics.rs]: `graph_root.span` is the span of the `lib.rs` file.  Using it to decide is definitely wrong.
The former version of this PR is the better middle ground.
Could you remove the last commit, and I'll approve the PR. Thanks

Comment (zbyrn) [reply compiler/rustc_resolve/src/late/diagnostics.rs:707]: Ah, my bad for misunderstanding what `graph_root` means! A reversion has been pushed.

Thanks for your comments and your approval!

**Comment by zbyrn** [comment]
@cjgillot I believe everything is ready

**Comment by cjgillot** [comment]
Could you squash the commits to remove the back-and-forth?
Thanks.

**Comment by zbyrn** [comment]
@cjgillot Done! Sorry for the confusion

**Pull request merged**